File: catalog/binary_search/ruby/binary_search.rb
# x: int, xs: int[] ordered asc
def contains(x, xs)
  if xs.length == 0 then
    false
  elsif xs.length == 1 then
    if xs[0] == x then
      true
    else
      false
    end
  else
    n = xs.length / 2
    if xs[n] == x then
      true
    elsif x < xs[n] then
      contains(x, xs.take(n))
    else
      contains(x, xs.drop(n))
    end
  end
end

File: catalog/binary_search/binary_search.hs
contains :: Ord a => a -> [a] -> Bool
contains _ [] = False
contains x [y]
    | x == y    = True
    | otherwise = False
contains x ys
    | x == m = True
    | x < m  = contains x (take p ys)
    | x > m  = contains x (drop p ys)
    where p = (length ys) `div` 2
          m = ys !! p

